### ISSUE TITLE
Add `lock` command and support `juv run Untitled.ipynb --locked`

### ISSUE THREAD

**Issue by manzt** [manzt/juv#7]
Ref: https://x.com/rahuldave/status/1848072186340196391

The idea would be to put a lockfile in the notebook metadata (either cell-level or top-level) and we could update the lock file on `run`/`add`/`lock`.

Then we could add a flag to `--locked` to skip dependency resolution and just create the enviroment from the lockfile.

**Comment by flying-sheep** [comment]
I saw this project and had the exact same thought!

In fact I don‘t think we’d need dependency specification in a code cell at all: jupyter has a perfect place for notebook-wide structured metadata already.

**Comment by manzt** [comment]
Should be able to benefit from:

- https://github.com/astral-sh/uv/issues/8609

When it lands to save the export to the notebook metadata.

**Comment by manzt** [comment]
> In fact I don‘t think we’d need dependency specification in a code cell at all: jupyter has a perfect place for notebook-wide structured metadata already.

Yeah, I'm a bit torn on this. Keeping in a notebook cell as a comment aligns closer with scripts, but hiding means we could do more to control the behavior of adding/removing dependencies. When/if we introduce locking, we could always write a new lock on any command (`add`, `run`, ...) and also offer something like `juv sync` to sync the lock.

**Comment by manzt** [comment]
uv scripts now also support `uv.tool.exclude-newer` metadata. Without a lock file, this could go a long way for making one-off notebooks more reproducible to add a timestamp to the notebook meta that we could have a flag to run.

**Comment by flying-sheep** [comment]
The purpose of putting it into the metadata isn’t supposed to hide it, but to allow us to

1. create a jupyterlab plugin to display and modify dependencies.
   since we start the jupyterlab instance with `juv run`, we’ll be able to inject the plugin into that environment, making that not a problem
2. move the feature from the plugin into jupyterlab upstream (possibly with some changes after feedback)
3. this makes it likely that alternative frontends like VS Code will also be able to support it

**Comment by tlambert03** [comment]
def, love the lock idea :)

**Comment by manzt** [comment]
See #64, uv now supports locking standalone scripts. 

```sh
uv lock --script foo.py # generates foo.py.lock
```

Right now I'm writing this metadata to the root metadata under a `"uv.lock"` key.

**Comment by manzt** [comment]
Released in [v0.3.0](https://github.com/manzt/juv/releases/tag/v0.3.0)!
